src/torch_constrained/constrained_optimizer.py

```diff
@@ -1,359 +1,289 @@
             primal_parameters = list(primal_parameters)
-
-        self.primal_optimizer = loss_optimizer(primal_parameters, lr_x, **primal_kwargs)
-        self.dual_optimizer_class = functools.partial(constraint_optimizer, lr=lr_y, **dual_kwargs)
-
-        self.dual_optimizer = None
-        self.augmented_lagrangian_coefficient = augmented_lagrangian_coefficient
-        self.alternating = alternating
-        self.shrinkage = shrinkage
-        self.dual_dtype = dual_dtype
+        self.primal_optimizer = primal_optim_class(primal_parameters, **primal_kwargs)
+        self.primal_parameters = primal_parameters
+
+        # Initialization values for the dual variables
+        self.dual_init_done = False
+        self.ineq_init = ineq_init
+        self.eq_init = eq_init
+
+        # Create flag for execution in un-constrained setting
+        self.is_constrained = dual_optim_class is not None 
+
+        if self.is_constrained:
+            # Important: the dual optimizer is instantiated in 'init_dual_variables'.
+            self.dual_optimizer_class = partial(dual_optim_class, **dual_kwargs)
+            self.dual_reset = dual_reset
+
+            # Other optimization and lagrangian options
+            self.augmented_lagrangian_coeff = augmented_lagrangian_coeff
+            self.alternating = alternating
+            self.shrinkage = shrinkage
+        else:
+            print('Unconstrained Execution')
+            if len(dual_kwargs) != 0:
+                warnings.warn("""Did not provide dual optimizer dual kwargs is non empty.
+                              Running with no dual optimizer and ignoring hyper-parameters.""")
+            self.dual_optimizer_class = None
+            self.dual_reset = False
+
+            # Other optimization and lagrangian options
+            self.augmented_lagrangian_coeff = 0.
+            self.alternating = False
+            self.shrinkage = 0.
+        
+        self.verbose = verbose 
 
         super().__init__(primal_parameters, {})
 
     def step(self, closure):
         def closure_with_shrinkage():
-            loss_, eq_defect_, inequality_defect_ = closure()
-            if self.shrinkage is not None and eq_defect_:
-                eq_defect_ = [self.shrinkage(e) for e in eq_defect_]
-
-            return loss_, eq_defect_, inequality_defect_
-
-        loss, eq_defect, inequality_defect = closure_with_shrinkage()
-
-        if not self.equality_multipliers and not self.inequality_multipliers:
-            self.init_dual_variables(eq_defect, inequality_defect, dtype=self.dual_dtype)
-
-        assert eq_defect is None or all([validate_defect(d, m) for d, m in zip(eq_defect, self.equality_multipliers)])
-        assert inequality_defect is None or all([d.shape == m.shape for d, m in zip(inequality_defect, self.inequality_multipliers)])
-
-        lagrangian = self.minmax_backward(loss, eq_defect, inequality_defect)
-
+            closure_dict = closure()
```

Comment:
returning dicts, is risky because it's hard to check their correctness at all time and specifying what should be inside and what should not be inside, also it's hard ensure the content doesn't change over time.
Jax solves this problem with namedtuples, I think that's a good tradeoff, additionally closure.equality is more readable than closure['equality']